mailbox/postgres/src/main/java/org/apache/james/mailbox/jpa/mail/dao/PostgresMailboxDAO.java

```diff
@@ -0,0 +1,165 @@
+import org.apache.james.mailbox.store.MailboxExpressionBackwardCompatibility;
+import org.jooq.DSLContext;
+import org.jooq.Record;
+import org.jooq.Record1;
+import org.jooq.exception.DataAccessException;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+public class PostgresMailboxDAO {
+    private static final char SQL_WILDCARD_CHAR = '%';
+    private static final String DUPLICATE_VIOLATION_MESSAGE = "duplicate key value violates unique constraint";
+
+    private final PostgresExecutor postgresExecutor;
+
+    @Inject
+    public PostgresMailboxDAO(PostgresExecutor postgresExecutor) {
+        this.postgresExecutor = postgresExecutor;
+    }
+
+    public Mono<Mailbox> create(MailboxPath mailboxPath, UidValidity uidValidity) {
+        final PostgresMailboxId mailboxId = PostgresMailboxId.generate();
+        return postgresExecutor.dslContext()
+            .flatMap(dslContext -> Mono.from(dslContext.insertInto(TABLE_NAME, MAILBOX_ID, MAILBOX_NAME, USER_NAME, MAILBOX_NAMESPACE, MAILBOX_UID_VALIDITY)
+                .values(mailboxId.asUuid(), mailboxPath.getName(), mailboxPath.getUser().asString(), mailboxPath.getNamespace(), uidValidity.asLong())))
+            .thenReturn(new Mailbox(mailboxPath, uidValidity, mailboxId))
+            .onErrorMap(e -> e instanceof DataAccessException && e.getMessage().contains(DUPLICATE_VIOLATION_MESSAGE),
+                e -> new MailboxExistsException(mailboxPath.getName()));
+    }
+
+    public Mono<MailboxId> rename(Mailbox mailbox) {
+        Preconditions.checkNotNull(mailbox.getMailboxId(), "A mailbox we want to rename should have a defined mailboxId");
+
+        return postgresExecutor.dslContext()
+            .flatMap(dslContext -> rename(mailbox, dslContext))
+            .onErrorMap(e -> e instanceof DataAccessException && e.getMessage().contains(DUPLICATE_VIOLATION_MESSAGE),
+                throwable -> new MailboxExistsException(mailbox.getName()));
+    }
+
+    private Mono<MailboxId> rename(Mailbox mailbox, DSLContext dslContext) {
+        return Flux.from(dslContext.update(TABLE_NAME)
+                .set(MAILBOX_NAME, mailbox.getName())
+                .set(USER_NAME, mailbox.getUser().asString())
+                .set(MAILBOX_NAMESPACE, mailbox.getNamespace())
+                .where(MAILBOX_ID.eq(((PostgresMailboxId) mailbox.getMailboxId()).asUuid()))
+                .returning(MAILBOX_ID))
+            .collect(ImmutableList.toImmutableList())
+            .flatMap(records -> {
+                if (records.size() == 0) {
+                    return Mono.error(new MailboxNotFoundException(mailbox.getMailboxId()));
+                } else {
+                    return Mono.just(mailbox.getMailboxId());
+                }
+            });
+    }
+
+    public Mono<Void> delete(MailboxId mailboxId) {
+        return postgresExecutor.dslContext()
```

Comment:
Some util method in executor class make code less boilerplate :D 
```
return postgresExecutor.executeVoid(dslContext -> Mono.from(dslContext.deleteFrom(TABLE_NAME)
                .where(MAILBOX_ID.eq(((PostgresMailboxId) mailboxId).asUuid()))));
```

Reply:
Is it necessary? I see that the amount of boilerplate it remove is very little

Reply:
Please do

Reply:
updated